### PR TITLE
tidb: New domain with retry and backoff

### PR DESCRIPTION
**Pull request by shenli**
@iamxy @coocood @tiancaiamao @zimulala PTAL
When start TiDB and TiKV at the same time, TiDB should sleep a while and try new domain again when get domain failed.

### CODE REVIEW

File: tidb.go [hunk 1]
```diff
@@ -65,7 +65,15 @@ func (dm *domainMap) Get(store kv.Storage) (d *domain.Domain, err error) {
 	if !localstore.IsLocalStore(store) {
 		lease = schemaLease
 	}
-	d, err = domain.NewDomain(store, lease)
+	for i := 1; i <= defaultMaxRetries; i++ {
```
Comment (ngaut) [inline tidb.go]: Can we implement some function RunWithRetry(retryCount, func() {}). It should be more readable.

**Comment by shenli** [comment]
@ngaut PTAL
@coocood @iamxy

File: util/misc.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+// The second is if the f meet any error.
+func RunWithRetry(retryCnt int, backoff uint64, f func() (bool, error)) (err error) {
+	for i := 0; i < retryCnt; i++ {
+		var retryAble bool
+		retryAble, err = f()
+		if err != nil {
+			if !retryAble {
+				break
+			}
+			sleepTime := time.Duration(backoff * uint64(i))
```
Comment (coocood) [inline util/misc.go]: The `backoff` is used ass nano second.

**Comment by shenli** [comment]
@coocood @ngaut PTAL

**Comment by coocood** [comment]
@shenli TiKV client use milliseconds as time unit, we better use the same unit.

**Comment by shenli** [comment]
@coocood Yes, will fix.

**Comment by shenli** [comment]
@coocood @tiancaiamao PTAL

File: tidb.go [hunk 3]
```diff
@@ -37,14 +37,15 @@ import (
...
 )
 
 // Engine prefix name
 const (
-	EngineGoLevelDBMemory = "memory://"
-	defaultMaxRetries     = 30
-	retrySleepInterval    = 500 * time.Millisecond
+	EngineGoLevelDBMemory        = "memory://"
+	defaultMaxRetries            = 30
+	retrySleepInterval    uint64 = 500
```
Comment (ngaut) [inline tidb.go]: s/retrySleepInterval/retryInterval/

File: util/misc.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+// RunWithRetry will run the f with backoff and retry.
+// retryCnt: Max retry count
+// backoff: When run f failed, it will sleep backoff * triedCount time.Millisecond.
+// Function f should have two return value. The first one is an bool which indicate if the err if retryable.
+// The second is if the f meet any error.
+func RunWithRetry(retryCnt int, backoff uint64, f func() (bool, error)) (err error) {
+	for i := 1; i <= retryCnt; i++ {
+		var retryAble bool
+		retryAble, err = f()
+		if err != nil {
```
Comment (ngaut) [inline util/misc.go]: if err == nil {
    return nil
}

//retry logical

**Comment by shenli** [comment]
PTAL @ngaut

**Comment by ngaut** [comment]
LGTM

**Comment by shenli** [comment]
@coocood PTAL

**Comment by tiancaiamao** [comment]
LGTM

**Comment by coocood** [comment]
LGTM

**Pull request merged**